README.md
# bash-messenger

## Description

Simple bash message routines for displaying consistent end user messages

![Screenshot](https://github.com/mikeerickson/bash-messenger/blob/master/docs/messenger-example.png)

## Installation

Using `curl` you can install the `bash-messenger` library

- move to global directory where you wish to install library (I typically use `~/scripts`)

- use the following `curl` command to install

    `curl -o messenger.sh https://raw.githubusercontent.com/mikeerickson/bash-messenger/master/src/messenger.sh`

## Usage

- Install where you wish to use with your bash scripts
  - I typically have a `~/scripts` directory where I reference all my utility and global bash scripts
- Source messenger library
    `source ~/scripts/messenger.sh`
- Wherever you wish to execute on of the messenger methods

    ```shell
      # standard message
      success "success message"

      # message with label
      success "success message" "SUCCESS"
    ```

### Messenger API

Messenger API includes a suite of utility methods which becomes available to any bash scripts by simply sourcing the `messenger.sh` file

```bash
source messenger.sh
```

#### Messenger Signature

All messenger methods use the same method signature as follows

| Parameter        | Type   | Description                                                                          |
| ---------------- | ------ | ------------------------------------------------------------------------------------ |
| message          | string | Desired message to display in console                                                |
| label [optional] | string | Optional label message, if supplied the label will appear first, followed by message |

#### Messenger Methods

**User Notification Messages**

```bash
error(msg:string, [label:string])

info(msg:string, [label:string])

log(msg:string, [label:string])

debug(msg:string, [label:string])

note(msg:string, [label:string])

notice(msg:string, [label:string])

status(msg:string, [label:string])

success(msg:string, [label:string])

warn(msg:string, [label:string])

warning(msg:string, [label:string])
```

**Utilities:**
```
  version() -> outputs module version
```

### License

Copyright &copy; 2019 Mike Erickson
Released under the MIT license

## Credits

bash-messenger written by Mike Erickson

E-Mail: [codedungeon@gmail.com](mailto:codedungeon@gmail.com)

Twitter: [@codedungeon](http://twitter.com/codedungeon)

Website: [codedungeon.io](http://codedungeon.io)

CHANGELOG.md
# <project_name> Changelog

## Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.1] - 2019-03-25

### Added

- Initial Release

[keep a changelog](https://keepachangelog.com/en/1.0.0/)


test.sh
#!/usr/bin/env bash

source src/messenger.sh
test_messenger

src/colors.sh
#!/usr/bin/env bash

# Colors used for status updates
ESC_SEQ="\x1b["
COL_RESET=${ESC_SEQ}"39;49;00m"
COL_RED=${ESC_SEQ}"0;31m"
COL_LRED=${ESC_SEQ}"31;01m"
COL_GREEN=${ESC_SEQ}"0;32m"
COL_LGREEN=${ESC_SEQ}"32;01m"
COL_YELLOW=${ESC_SEQ}"0;33m"
COL_LYELLOW=${ESC_SEQ}"33;01m"
COL_BLUE=${ESC_SEQ}"0;34m"
COL_LBLUE=${ESC_SEQ}"34;01m"
COL_MAGENTA=${ESC_SEQ}"0;35m"
COL_LMAGENTA=${ESC_SEQ}"35;01m"
COL_CYAN=${ESC_SEQ}"0;36m"
COL_LCYAN=${ESC_SEQ}"36;01m"
COL_WHITE=${ESC_SEQ}"0;37m"
COL_WHITE=${ESC_SEQ}"37;01m"

# added COLOR... prefix, leaving COL... for backwards compatability
COLOR_RESET=${ESC_SEQ}"39;49;00m"
COLOR_RED=${ESC_SEQ}"0;31m"
COLOR_LRED=${ESC_SEQ}"31;01m"
COLOR_GREEN=${ESC_SEQ}"0;32m"
COLOR_LGREEN=${ESC_SEQ}"32;01m"
COLOR_YELLOW=${ESC_SEQ}"0;33m"
COLOR_LYELLOW=${ESC_SEQ}"33;01m"
COLOR_BLUE=${ESC_SEQ}"0;34m"
COLOR_LBLUE=${ESC_SEQ}"34;01m"
COLOR_MAGENTA=${ESC_SEQ}"0;35m"
COLOR_LMAGENTA=${ESC_SEQ}"35;01m"
COLOR_CYAN=${ESC_SEQ}"0;36m"
COLOR_LCYAN=${ESC_SEQ}"36;01m"
COLOR_WHITE=${ESC_SEQ}"0;37m"
COLOR_WHITE=${ESC_SEQ}"37;01m"


### SAMPLE
# printf "\n ${COL_GREEN}[ •• ]${COL_RESET} ${COL_YELLOW} ==> Copying files to ${COL_CYAN}./examples${COL_GREEN} ...   ${COL_RESET}\n"
###


src/messenger.sh
#!/usr/bin/env bash
# messenger.sh

VERSION="1.0.0"

ESC_SEQ="\x1b["
COL_RESET=${ESC_SEQ}"39;49;00m"
COL_GREEN=${ESC_SEQ}"0;32m"
COL_RED=${ESC_SEQ}"0;31m"
COL_YELLOW=${ESC_SEQ}"0;33m"
COL_BLUE=${ESC_SEQ}"0;34m"
COL_MAGENTA=${ESC_SEQ}"0;35m"
COL_CYAN=${ESC_SEQ}"0;36m"
COL_WHITE=${ESC_SEQ}"0;37m"

test_messenger() {
    printf "\n"
    
    version
    printf "\n"
    printf "\n"
    
    success "Success Message" " SUCCESS "
    success "Success Message"
    printf "\n"
    
    error "Error Message" " ERROR "
    error "Error Message"
    printf "\n"
    
    warning "Warning Message" " WARNING "
    warning "Warning Message"
    printf "\n"
    
    warn "Warn Message" " WARN "
    warn "Warn Message"
    printf "\n"
    
    info "Info Message" " INFO "
    info "Info Message"
    printf "\n"
    
    note "Note Message" " NOTE "
    note "Note Message"
    printf "\n"
    
    log "Log Message" " LOG "
    log "Log Message"
    printf "\n"
    
    debug "Debug Message" " DEBUG "
    debug "Debug Message"
    printf "\n"
    
    status "Status Message" " STATUS "
    status "Status Message"
    printf "\n"
    
    important "Important Message" " IMPORTANT "
    important "Important Message"
    printf "\n"
    
    notice "Notice Message" " NOTICE "
    notice "Notice Message"
    printf "\n"
}

version() {
    info "Bash Messenger v$VERSION"
    warning "Crafted with love by Codedungeon"
}

success() {
    if [ -z "$2" ]; then
        printf "$COL_GREEN$1$COL_RESET\n";
    else
        printf "\e[30;42;11;87m$2$COL_GREEN $1 $COL_RESET\n";
    fi
}

error() {
    if [ -z "$2" ]; then
        printf "$COL_RED$1$COL_RESET\n";
    else
        printf "\e[30;41;11;87m$2$COL_RED $1$COL_RESET\n";
    fi
}

info() {
    if [ -z "$2" ]; then
        printf "$COL_BLUE$1$COL_RESET\n";
    else
        printf "\e[30;44;11;87m$2$COL_BLUE $1$COL_RESET\n";
    fi
}

warning() {
    if [ -z "$2" ]; then
        printf "$COL_YELLOW$1$COL_RESET\n";
    else
        printf "\e[30;43;11;87m$2$COL_YELLOW $1$COL_RESET\n";
    fi
}

warn() {
    if [ -z "$2" ]; then
        printf "$COL_YELLOW$1$COL_RESET\n";
    else
        printf "\e[30;43;11;87m$2$COL_YELLOW $1$COL_RESET\n";
    fi
}

important() {
    if [ -z "$2" ]; then
        printf "\e[33m$1$COL_RESET\n";
    else
        printf "\e[30;43;11;87m$2$COL_RESET\e[33m $1$COL_RESET\n";
    fi
}

log() {
    if [ -z "$2" ]; then
        printf "$1\n";
    else
        printf "\e[30;107;11;87m$2$COL_RESET $1\n";
    fi
}

debug() {
    if [ -z "$2" ]; then
        printf "$1\n";
    else
        printf "\e[30;107;11;87m$2$COL_RESET $1\n";
    fi
}

status() {
    if [ -z "$2" ]; then
        printf "$COL_MAGENTA$1$COL_RESET\n";
    else
        printf "\e[30;45;11;87m$2$COL_MAGENTA $1$COL_RESET\n";
    fi
}

note() {
    if [ -z "$2" ]; then
        printf "$1\n";
    else
        printf "\e[30;107;11;87m$2$COL_RESET $1\n";
    fi
}

notice() {
    if [ -z "$2" ]; then
        printf "$COL_CYAN$1$COL_RESET\n";
    else
        printf "\e[30;46;11;87m$2$COL_CYAN $1$COL_RESET\n";
    fi
}
